### PR TITLE
[rpc] remove default constructor in futureInfo

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30197 [rpc] remove default constructor in futureInfo**

Addresses @pietern's comment on https://github.com/pytorch/pytorch/pull/29601. This default constructor was added because `std::map`'s `operator[]` requires a default constructor. However, instead of using `operator[]`, we can
use `emplace` and remove the constructor, to ensure that the `FutureInfo` struct
doesn't get constructed with garbage values.

Differential Revision: [D18627675](https://our.internmc.facebook.com/intern/diff/D18627675/)

### CODE REVIEW

File: torch/csrc/distributed/rpc/process_group_agent.h [hunk 1]
```diff
@@ -93,7 +93,6 @@ class ProcessGroupAgent : public RpcAgent {
           startTime_(startTime),
           dstRank_(dstRank),
           timeout_(timeout) {}
-    FutureInfo() {}
```
Comment (xush6528) [inline torch/csrc/distributed/rpc/process_group_agent.h]: Delete explicitly?
```cpp
FutureInfo() = delete;
```

Comment (xush6528) [inline torch/csrc/distributed/rpc/process_group_agent.h]:
```cpp
FutureInfo() = delete;
```

**Comment by zhaojuanmao** [comment]
@rohan-varma  I just caught up diff review about future time out PR in #29601, I think we do not need startTime_ and timeout_, endTime_ should be sufficient.

**Comment by pietern** [comment]
Thanks, @rohan-varma. This makes it just a bit more robust for future changes.

**Comment by xush6528** [comment]
@zhaojuanmao 

I am doing it in https://github.com/pytorch/pytorch/pull/30342.

**Review by pietern — APPROVED** [review]
LGTM, thank you!

**Pull request closed without merging**